### PR TITLE
SummaryWriter.add_figure: add type hints

### PR DESCRIPTION
**Pull request by adamjstewart**
Discovered a bug in our code that could have been prevented by type hints, so I added them 😄

### CODE REVIEW

File: torch/utils/tensorboard/writer.py [hunk 1]
```diff
@@ -3,9 +3,11 @@
 
 import os
 import time
+from typing import List, Optional, Union
 
 import torch
 
+from matplotlib.figure import Figure
```
Comment (Skylion007) [inline torch/utils/tensorboard/writer.py:10]: I don't think this is a hard dependency in Pytorch is it? If not we probably need to hide it behind a TYPE_CHECKING if.

**Comment by adamjstewart** [comment]
@Skylion007 any other required changes?

**Comment by adamjstewart** [comment]
@malfet or anyone else that can review?

**Comment by adamjstewart** [comment]
@ezyang?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**